include/alpaka/core/ApiCudaRt.hpp

```diff
@@ -0,0 +1,377 @@
+
+        // Host function helper
+        // Encapsulates the different function signatures used by cudaStreamAddCallback and cudaLaunchHostFn, and the
+        // different calling conventions used by CUDA (__stdcall on Win32) and HIP (standard).
+        struct HostFnAdaptor
+        {
+            HostFn_t func_;
+            void* data_;
+
+            static void CUDART_CB hostFunction(void* data)
+            {
+                auto ptr = reinterpret_cast<HostFnAdaptor*>(data);
+                ptr->func_(ptr->data_);
+                delete ptr;
+            }
+
+            static void CUDART_CB streamCallback(Stream_t, Error_t, void* data)
+            {
+                auto ptr = reinterpret_cast<HostFnAdaptor*>(data);
+                ptr->func_(ptr->data_);
+                delete ptr;
+            }
+        };
+
+        // Runtime API
+        static inline Error_t deviceGetAttribute(int* value, DeviceAttr_t attr, int device)
+        {
+            return ::cudaDeviceGetAttribute(value, attr, device);
+        }
+
+        static inline Error_t deviceGetLimit(size_t* pValue, Limit_t limit)
+        {
+            return ::cudaDeviceGetLimit(pValue, limit);
+        }
+
+        static inline Error_t deviceReset()
+        {
+            return ::cudaDeviceReset();
+        }
+
+        static inline Error_t deviceSetLimit(Limit_t limit, size_t value)
+        {
+            return ::cudaDeviceSetLimit(limit, value);
+        }
+
+        static inline Error_t deviceSynchronize()
+        {
+            return ::cudaDeviceSynchronize();
+        }
+
+        static inline Error_t eventCreate(Event_t* event)
+        {
+            return ::cudaEventCreate(event);
+        }
+
+        static inline Error_t eventCreateWithFlags(Event_t* event, unsigned int flags)
+        {
+            return ::cudaEventCreateWithFlags(event, flags);
+        }
+
+        static inline Error_t eventDestroy(Event_t event)
+        {
+            return ::cudaEventDestroy(event);
+        }
+
+        static inline Error_t eventQuery(Event_t event)
+        {
+            return ::cudaEventQuery(event);
+        }
+
+        static inline Error_t eventRecord(Event_t event, Stream_t stream)
+        {
+            return ::cudaEventRecord(event, stream);
+        }
+
+        static inline Error_t eventSynchronize(Event_t event)
+        {
+            return ::cudaEventSynchronize(event);
+        }
+
+        static inline Error_t free(void* devPtr)
+        {
+            return ::cudaFree(devPtr);
+        }
+
+        static inline Error_t freeAsync(void* devPtr, Stream_t stream)
+        {
+#    if BOOST_LANG_CUDA >= BOOST_VERSION_NUMBER(11, 2, 0)
+            return ::cudaFreeAsync(devPtr, stream);
+#    else
+            // Unused.
+            static_cast<void>(devPtr);
+            static_cast<void>(stream);
+            // Not implemented.
+            return errorUnknown;
+#    endif
+        }
```

Comment:
I see... how would you see this approach instead ?
```suggestion
#    if BOOST_LANG_CUDA >= BOOST_VERSION_NUMBER(11, 2, 0)
        static inline Error_t freeAsync(void* devPtr, Stream_t stream)
        {
            return ::cudaFreeAsync(devPtr, stream);
        }
#    else
        static inline Error_t freeAsync(void*, Stream_t)
        {
            // Not implemented.
            return errorUnknown;
#    endif
        }
```

Reply:
@j-stephan

Reply:
In general I'm fine with using `[[maybe_unused]]`, for example when the conditions when a parameter is used or not are not obvious.

In this specific case we know when the parameters are used or not, so I though this approach would make more sense.

But, all in all, I don't mind much, and I'm fine using whatever you prefer.

Reply:
> In this specific case we know when the parameters are used or not, so I though this approach would make more sense.

Sure, but `[[maybe_unused]]` is a compiler hint ;-) We also know this when using `if constexpr`, for example. For that reason I prefer `[[maybe_unused]]`.

Reply:
OK.